File: CPP/Traversal.cpp
#include <cstdio>
#include <vector>
using namespace std;

vector<int> postOrder;

int indexOf(int value, vector<int> in) {
	for (int i = 0; i < in.size(); i++) {
		if (value == in[i]) return i;
	}
	return -1;
}

template <typename T>
vector<T> subList(vector<T> arr, int begin, int end) {
	return vector<T>(arr.begin() + begin, arr.begin() + end);
}

void printPostOrder(vector<int> pre, vector<int> in) {
	if (pre.empty()) return;
	int rootValue = pre[0];
	int rootIndex = indexOf(rootValue, in);
	printPostOrder(subList(pre, 1, rootIndex + 1), subList(in, 0, rootIndex));
	printPostOrder(subList(pre, rootIndex + 1, pre.size()), subList(in, rootIndex + 1, in.size()));
	postOrder.push_back(rootValue);

}

int main() {
	int cases; scanf("%d", &cases);
	while (cases--) {
		int size; scanf("%d", &size);
		vector<int> pre, in;
		for (int i = 0; i < size; i++) {
			int temp; scanf("%d", &temp); pre.push_back(temp);
		}
		for (int i = 0; i < size; i++) {
			int temp; scanf("%d", &temp); in.push_back(temp);
		}
		printPostOrder(pre, in);
		for (int i = 0; i < postOrder.size(); i++) {
			if (i != 0) printf(" ");
			printf("%d", postOrder[i]);
		}
		printf("\n");
		postOrder.clear();
	}
}

File: JAVA/Traversal.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.List;


public class Traversal {
	
	private static ArrayList<Integer> postOrder;
	
	public static void main(String[] args) {
		try {
			BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
			BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
			int cases = Integer.parseInt(reader.readLine());
			
			while (cases-- > 0) {
				int n = Integer.parseInt(reader.readLine());
				ArrayList<Integer> pre = new ArrayList<Integer>(), in = new ArrayList<Integer>();
				postOrder = new ArrayList<Integer>();
				String[] pre_input = reader.readLine().split(" "), in_input = reader.readLine().split(" ");
				for (int i = 0; i < n; i++) {
					pre.add(Integer.parseInt(pre_input[i])); in.add(Integer.parseInt(in_input[i]));
				}
				printPostOrder(pre, in);
				for (int i = 0; i < postOrder.size(); i++) {
					if (i != 0) writer.append(" ");
					writer.append(String.valueOf(postOrder.get(i)));
				}
				writer.newLine();
			}
			writer.flush();
			writer.close(); reader.close();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	public static void printPostOrder(List<Integer> pre, List<Integer> in) {
		if (pre.size() == 0) return ;
		int rootValue = pre.get(0);
		int rootIndex = in.indexOf(rootValue);
		printPostOrder(pre.subList(1, rootIndex + 1), in.subList(0, rootIndex));
		printPostOrder(pre.subList(rootIndex + 1, pre.size()), in.subList(rootIndex + 1,  in.size()));
		postOrder.add(rootValue);
	}
}
